lib/ace/mode/lua.js

```diff
@@ -77,6 +77,22 @@ oop.inherits(Mode, TextMode);
             if (token.type == "keyword") {
                 if (token.value in indentKeywords) {
                     level += indentKeywords[token.value];
+                    if (/\b(function|then|do)\b|{\s*$|(\[=*\[)/.test(token.value)){
+                        setTimeout(() => {
+                            let size = editor.getOption("tabSize");
+                            let position = editor.getCursorPosition();
+                            editor.session.insert(position, "\n");
+                            let text = editor.session.getLine(position.row);
+                            text = text.split(" ".repeat(size));
+                            if (text.length == size/2){
+                                editor.session.insert(editor.getCursorPosition(), "end");
+                                editor.moveCursorTo(position.row, position.column);
+                            } else {
+                                editor.session.insert(editor.getCursorPosition(), (" ".repeat(size)).repeat(text.length-2) + "end");
+                                editor.moveCursorTo(position.row, position.column);
+                            }
+                        }, 5);
```

Comment:
1. why is timeout needed
2. should this be added every time when inserting new line or only the first time after wring the keyword?

Reply:
timeout is needed because the line is considered an empty string for a few milliseconds.
its only added everytime the regex matches a statement for it, e.g. do, then, function, elseif.